lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -0,0 +1,251 @@
+    }
+    return true;
+  }
+
+  void runOnOperation() override {
+    auto &instanceGraph = getAnalysis<InstanceGraph>();
+    // Get the post order traversal of the modules in the InstanceGraph.
+    SmallVector<FModuleOp, 0> modules(
+        llvm::map_range(llvm::post_order(&instanceGraph), [](auto *node) {
+          return cast<FModuleOp>(*node->getModule());
+        }));
+    // Traverse the modules in post order.
+    for (auto module : modules) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\n Traversing module:" << module.moduleNameAttr());
+      for (Operation &op :
+           llvm::make_early_inc_range(module.getBodyBlock()->getOperations()))
+        // This only handles the RefSendOp (and RefRecvOp when its implemented)
+        if (handleRemoteReference(&op, module))
+          continue;
+        else if (handleInstanceOp(&op, module))
+          continue;
+
+      for (size_t portNum = 0, e = module.getNumPorts(); portNum < e; ++portNum)
+        if (module.getPortType(portNum).isa<RefType>()) {
+          refPortsToRemoveMap[module].push_back(portNum);
+        }
+    }
+    for (auto op : opsToRemove)
+      op->erase();
+    for (auto iter : refPortsToRemoveMap)
+      if (auto mod = dyn_cast<FModuleOp>(iter.getFirst()))
+        mod.erasePorts(iter.getSecond());
+      else if (auto inst = dyn_cast<InstanceOp>(iter.getFirst())) {
+        ImplicitLocOpBuilder b(inst.getLoc(), inst);
+        inst.erasePorts(b, iter.getSecond());
+        inst.erase();
+      }
+  }
+
+  /// Get the cached namespace for a module.
+  ModuleNamespace &getModuleNamespace(FModuleLike module) {
+    auto it = moduleNamespaces.find(module);
+    if (it != moduleNamespaces.end())
+      return it->second;
+    return moduleNamespaces.insert({module, ModuleNamespace(module)})
+        .first->second;
+  }
+
+  InnerRefAttr getInnerRefTo(Operation *op) {
+    return ::getInnerRefTo(op, "xmr_sym",
+                           [&](FModuleOp mod) -> ModuleNamespace & {
+                             return getModuleNamespace(mod);
+                           });
+  }
+  /// Cached module namespaces.
+  DenseMap<Operation *, ModuleNamespace> moduleNamespaces;
+
+  /// Map of a ref port to the RefSend op. Each ref port can be satically
+  /// resolved to a single remote send op, according to the constraints on the
+  /// ref ports. The pair represents the module and the port number.
+  DenseMap<std::pair<Operation *, size_t>, Operation *> refPortToXmrMap;
+  /// Map of the remote RefSend op, to the sequence of the InnerRef symbols that
+  /// represents the Final XMR. Each of the intermediary symbol is an instance
+  /// op, and the final symbol represents the leaf operation corresponding to
+  /// the XMR. The symbol sequence can represent a hierarchical path to the
+  /// operation, for which the XMR needs to be constructed.
+  DenseMap<Operation *, SmallVector<Attribute>> remoteRefToXMRsymVec;
+  SmallVector<Operation *> opsToRemove;
```

Comment:
Can `opsToRemove` be entirely avoided by deleting ops as you go? If that works, then the only things that need to be deleted after the fact are to iterate over the `remoteRefToXMRsymVec` keys.